services/worker/src/worker/job_runners/config/parquet_and_info.py

```diff
@@ -0,0 +1,910 @@
+            If we failed to get the external files sizes to make sure we can convert the dataset to parquet
+        - [`~job_runners.config.parquet_and_info.ExternalFilesSizeRequestTimeoutError`]
+            If we failed to get the external files sizes to make sure we can convert the dataset to parquet
+        - [`~job_runners.config.parquet_and_info.ExternalFilesSizeRequestError`]
+            If we failed to get the external files sizes to make sure we can convert the dataset to parquet
+    </Tip>
+    """
+    logging.info(f"get parquet files and dataset info for {dataset=} {config=}")
+
+    raise_if_not_supported(
+        dataset=dataset,
+        config=config,
+        hf_endpoint=hf_endpoint,
+        hf_token=hf_token,
+        committer_hf_token=committer_hf_token,
+        revision=source_revision,
+        supported_datasets=supported_datasets,
+        blocked_datasets=blocked_datasets,
+        max_dataset_size=max_dataset_size,
+    )
+
+    hf_api = HfApi(endpoint=hf_endpoint, token=hf_token)
+    committer_hf_api = HfApi(endpoint=hf_endpoint, token=committer_hf_token)
+
+    # prepare the parquet files locally
+    parquet_files: List[ParquetFile] = []
+    download_config = DownloadConfig(delete_extracted=True)
+    try:
+        builder = load_dataset_builder(
+            path=dataset,
+            name=config,
+            revision=source_revision,
+            use_auth_token=hf_token,
+            download_config=download_config,
+        )
+    except _EmptyDatasetError as err:
+        raise EmptyDatasetError(f"{dataset=} is empty.", cause=err) from err
+    raise_if_too_big_from_external_data_files(
+        builder=builder,
+        max_dataset_size=max_dataset_size,
+        max_external_data_files=max_external_data_files,
+        hf_token=hf_token,
+    )
+    builder.download_and_prepare(file_format="parquet")  # the parquet files are stored in the cache dir
+    dataset_info = asdict(builder.info)
+    # ^ see
+    # https://github.dev/huggingface/datasets/blob/e183a269067575db8765ee979bd8523d14a1adae/src/datasets/info.py#L244-L245
+    # note that asdict() is not typed in the datasets library, hence type: ignore
+    parquet_files.extend(
+        ParquetFile(local_file=local_file, local_dir=builder.cache_dir, config=config)
+        for local_file in glob.glob(f"{builder.cache_dir}**/*.parquet")
+    )
+
+    # create the target revision if it does not exist yet
+    try:
+        refs = hf_api.list_repo_refs(repo_id=dataset, repo_type=DATASET_TYPE)
+        if all(ref.ref != target_revision for ref in refs.converts):
+            committer_hf_api.create_branch(
+                repo_id=dataset, branch=target_revision, repo_type=DATASET_TYPE, revision=source_revision
+            )
+    except RepositoryNotFoundError as err:
+        raise DatasetNotFoundError("The dataset does not exist on the Hub.") from err
+
+    # delete:
```

Comment:
Here, working at config level, instead of dataset level, means that we need to work in the config directory, not in the dataset repository (every config stores the parquet files in its own directory).
If we delete all the files here, we will never have more than one config at the same time. We have to let all the other config directories untouched, and only clear the config's own directory.

Which also leads to an issue I hadn't thought about before: which step should take care of deleting obsolete config directories? It's surely a rare use case, but if a commit in the repository deletes a config (for example, the dataset had 4 configs, and now it has only 3): which step should delete the directory of the removed config?

Reply:
> hf api should care for this, no?

yes: see https://github.com/huggingface/datasets-server/blob/main/services/worker/src/worker/job_runners/parquet_and_dataset_info.py#L840 (https://huggingface.co/docs/huggingface_hub/package_reference/hf_api#huggingface_hub.HfApi.create_commit.parent_commit)

> Specifying parent_commit ensures the repo has not changed before committing the changes, and can be especially useful if the repo is updated / committed to concurrently.

Reply:
> alternatively this can be performed in /config-names, because when the set of config is changed, the processing graph is triggered, right? and in this case there shouldn't be any problems with parallel deletion I guess. so maybe this option is better.

I think it's better to let the "parquet-and-info" step in charge of the parquet files. BTW: storing them on the Hub is an implementation detail, we could have stored them elsewhere (on S3).